Sources/Components/CarouselView/CarouselView.swift

```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    public weak var delegate: CarouselViewDelegate?
+
+    // MARK: - Private Properties
+
+    private var initialLayout = true
+
+    private lazy var layout: UICollectionViewFlowLayout = {
+        let layout = UICollectionViewFlowLayout()
+        layout.minimumLineSpacing = 0
+        layout.minimumInteritemSpacing = 0
+        layout.scrollDirection = .horizontal
+        return layout
+    }()
+
+    private lazy var collectionView: UICollectionView = {
+        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        collectionView.isPagingEnabled = true
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        return collectionView
+    }()
+
+    // MARK: - Init
+
+    public override init(frame: CGRect) {
+        super.init(frame: frame)
+        setup()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Overrides
+
+    public override var backgroundColor: UIColor? {
+        didSet {
+            collectionView.backgroundColor = backgroundColor
+        }
+    }
+
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+
+        guard initialLayout else { return }
+        initialLayout = false
+
+        // Scroll to first item
+        collectionView.contentOffset = CGPoint(
+            x: bounds.width,
+            y: collectionView.contentOffset.y
+        )
+    }
+}
+
+public extension CarouselView {
+    var contentOffset: CGPoint {
+        return CGPoint(
+            x: collectionView.contentOffset.x - collectionView.bounds.width,
+            y: collectionView.contentOffset.y
+        )
+    }
+
+    func register(_ cellClass: CarouselViewCell.Type) {
+        collectionView.register(cellClass)
+    }
+
+    func dequeue<T>(_ cellType: T.Type, for indexPath: IndexPath) -> T where T: CarouselViewCell {
+        return collectionView.dequeue(cellType, for: indexPath)
+    }
+
+    func scrollToItem(at indexPath: IndexPath, at scrollPosition: UICollectionView.ScrollPosition, animated: Bool) {
+        collectionView.scrollToItem(at: IndexPath(item: indexPath.item + 1, section: 0), at: scrollPosition, animated: animated)
+    }
+}
+
+extension CarouselView: UICollectionViewDataSource {
+    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let dataSource = dataSource else {
+            return 0
+        }
+
+        return dataSource.numberOfItems(in: self) + 2
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let dataSource = dataSource else {
+            preconditionFailure("Data source is not available")
```

Comment:
We could avoid doing this by having `dataSource` as a parameter in `init` method.

Reply:
`dataSource` would still be optional though. What if we do `collectionView.dataSource = self // or nil` in the `didSet` method of `dataSource`? Then this would never occur.

Reply:
What if we make it `private weak var dataSource: CarouselViewDataSource?`, so you can inject it only during initialization of `CarouselView`.

Reply:
Yep, that sounds good